File: swift/GameModel.swift
//
//  Model.swift
//  maze-escape
//
//  Created by Wayne Howarth on 24/07/2022.
//  Copyright © 2022 Wayne Howarth. All rights reserved.
//

import Foundation

class GameModel: CharacterModelListener {
	
	private var width: Int
	private var height: Int
	private var currentScore: Int
	private var heroCharacter: CharacterModel?
	
	private var listener: GameModelListener?
	private var generator: MazeGenerator!
	
	private var maze: Maze
	
	var score: Int {
		get {
			return currentScore
		}
	}
	
	public init(width: Int, height: Int, generator: MazeGenerator) throws {
		self.width = width
		self.height = height
		self.generator = generator
		self.currentScore = 0
		self.heroCharacter = nil
		
		maze = generator.generateMaze()
	}
	
	public func registerListener(listener: GameModelListener) {
		self.listener = listener
		
		listener.setMazeDimensions(width: width, height: height)
		heroCharacter = listener.createHeroCharacter()
		heroCharacter?.registerListener(listener: self)
		listener.draw(heroCharacter!)
	}
	
	public func newGame() {
		currentScore = 0
		maze = generator.generateMaze()
	}
	
	public func moveForward() {
		heroCharacter?.moveForward()
	}
	
	public func turnRight() {
		heroCharacter?.turnRight()

		if let listener = listener {
			listener.turnRight()
		}
	}
	
	public func turnLeft() {
		heroCharacter?.turnLeft()
		
		if let listener = listener {
			listener.turnLeft()
		}
	}
	
	func canMove(_ direction: Direction, from: (Int, Int)) -> Bool {
		
		switch (direction)
		{
			case .North:
				return maze.canMoveNorth(from: from)
			case .East:
				return maze.canMoveEast(from: from)
			case .South:
				return maze.canMoveSouth(from: from)
			case .West:
				return maze.canMoveWest(from: from)
			case.None:
				return false
		}
	}
	
	func moved(_ character: CharacterModel) {
		if let listener = listener {
			listener.draw(character)
		}
	}
}

File: scripts/GameModel.ts
class GameModel implements ICharacterModelListener {
    private _score: number;
    private _generator: IMazeGenerator;
    private _listener: IGameModelListener;
    private _maze: Maze;
    private _hero: CharacterModel;

    constructor(generator: IMazeGenerator) {
        this._score = 0;
        this._generator = generator;
        this._maze = generator.generateMaze();
    }

    getScore(): number {
        return this._score;
    }

    registerListener(listener: IGameModelListener) {
        this._listener = listener;

        this._listener.setMazeDimensions(this._maze.width(), this._maze.height());
        this._hero = new CharacterModel(this._maze.width() / 2, 0);
        this._hero?.registerListener(this);
        listener.draw(this._hero);
    }

    newGame() {
        this._score = 0;
        this._maze = this._generator.generateMaze();
    }

    moveforward() {
        this._hero?.moveForward();
    }

    turnRight() {
        this._hero?.turnRight();
        this._listener?.turnRight();
    }

    turnLeft() {
        this._hero?.turnLeft();
        this._listener?.turnLeft();
    }

    canMove(direction: Direction, x: number, y: number): boolean {
        switch (direction) {
            case Direction.North:
                return this._maze.canMoveNorth(x, y);
            case Direction.South:
                return this._maze.canMoveSouth(x, y);
            case Direction.East:
                return this._maze.canMoveEast(x, y);
            case Direction.West:
                return this._maze.canMoveWest(x, y);
            case Direction.None:
                return false;
        }
    }

    moved(character: CharacterModel) {
        this._listener?.draw(character);
    }
}